File: AlgorithmSeminar/15961.cpp
#include <bits/stdc++.h>
using namespace std;

int n, d, k, c;
vector<int> sushi;
map<int, int> m;


int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	cin >> n >> d >> k >> c;
	for (int i = 0,t; i < n; i++) {
		cin >> t; sushi.push_back(t);
		if (m.find(t) == m.end())m.insert(make_pair(t, 0));
	}
	if (m.find(c) == m.end())m.insert(make_pair(c, 0));

	int left = 0, right = 0, cnt = 0, kind = 0, answer = 0;
	while (left != n - 1) {
		if (cnt >= k) {
			if (--m[sushi[left]] == 0)
				kind--;
			left = (left + 1) % n;
			cnt--;
		}
		else {
			if (m[sushi[right]]++ == 0)
				kind++;
			right = (right + 1) % n;
			cnt++;
		}
		if (kind >= answer) {
			answer = kind;
			if (m[c] == 0)answer++;
		}
	}
	cout << answer;
	return 0;
}

File: python/15961.py
import sys

sys.stdin=open('input.txt','r')
n,d,k,c=map(int,sys.stdin.readline().split())
sushi=[]
for _ in range(n):
    sushi.append(int(sys.stdin.readline()))
keys=dict.fromkeys(set(sushi),0)
if keys.get(c) is None:keys[c]=0

left=0
right=0
cnt=0
kind=0
answer=0
while left!=n-1:
    #print(str(left)+" "+str(right))
    if cnt>=k:
        keys[sushi[left]]-=1
        if keys[sushi[left]]==0:
            kind-=1

        left=(left+1)%n
        cnt-=1
    else:
        if keys[sushi[right]]==0:
            kind+=1
        keys[sushi[right]]+=1
        right=(right+1)%n
        cnt+=1
    if kind>=answer:
        answer=kind
        if keys[c]==0:answer+=1
#print(sushi)
print(answer)